managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/PositionInfoUtils.java

```diff
@@ -0,0 +1,1762 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.bookkeeper.mledger.impl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.function.Consumer;
+import org.apache.bookkeeper.mledger.Position;
+import org.apache.pulsar.common.allocator.PulsarByteBufAllocator;
+
+final class PositionInfoUtils {
```

Comment:
I request to avoid creating Util Class for Position object as I have seen Position casting into client or in other system eg: Apache Beam connector and Once we have Util class, we are opening can of warm. Instead if we have an interface of CursorRecovery then this entire logic can go into `MultiEntryCursorRecoveryImplementation` Class,.

Reply:
PositionInfoUtils addresses problem that appears when serializing large PositionInfo. IIRC LightProto hogs the memory allocated for larger repeated objects and when the size of PositionInfo drop the memory isn't freed. 
This is used just for the serialization in the same compatible way that can be read by protobuf and doesn't create problems for any system that relies on proto definitions.

Reply:
is PositionInfoUtils auto generated?

Reply:
it is generated by light proto with serializePositionInfo() crafted manually to avoid some of the inefficiencies. Unfortunately, the generated code couldn't be (re)used directly as some of the required things are hidden with private access.

Reply:
oh.. this seems like a big red flag as we are combining auto generated code with actual code. this can make a nightmare when we will have a change in auto-generated code and we need to merge it with actual code. this is a huge red flag of this PR and we should avoid it.

Reply:
This is some sort of trick, but there is no better option at the moment, Protobuf is not able to handle efficiently serialization/deserialization while Lightproto is great.

Unfortunately there is not a good way to let the LightProto compiler fully create this file at build time, as we (I worked on the original patch) had to amend the generated code.

All the tricks are in this file, and not in other places.

As @dlg99 mentioned introducing some interface/abstraction will make the change more complicated.

Reply:
But I would avoid adding such a hack for just this feature in Pulsar. This shows that we should make interface for unack-message store/recovery and make it pluggable so, anyone can add pluggable solution and add a hack if needed. because by just looking at the amount of code, complication and hack for this feature in this PR , it's a great example of making it a pluggable and injecting custom code to keep Pulsar project free from hacks.